### PR TITLE
Make saving an order robust against email template errors

### PR DESCRIPTION
**Pull request by windaishi**
This PR is strongly related to https://github.com/shopware/shopware/pull/1496 and depends on it to be merged first.

### 1. Why is this change necessary?
When changing the status of an order to a status with an email template and this email template is corrupted, saving the order fails hard with an exception. The order gets saved correctly but no `after`-hooks are executed. Together with Shopware ERP (Powered by Pickware) this leads to serious stock errors that are hard to identify and recover from.

### 2. What does this change do, exactly?
It improves the `Order/save` action so it catches any exceptions that occur during email generation and reports the problem to the backend user.

### 3. Describe each step to reproduce the issue or behaviour.
1. Corrupt a status mail template, for example by adding an `{if}` tag without a corresponding closing `{/if}` tag.
2. Change the status of an order that would send this mail template on status change.

Result after applying this changeset:
image

### 4. Please link to the relevant issues (if any).
N/A

### 5. Which documentation changes (if any) need to be made because of this PR?
N/A

### 6. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: engine/Shopware/Controllers/Backend/Order.php [hunk 1]
```diff
@@ -716,6 +736,11 @@ public function batchProcessAction()
             try {
                 Shopware()->Models()->flush($order);
             } catch (Exception $e) {
+                $data['success'] = false;
+                $data['errorMessage'] = sprintf(
+                    $namespace->get('save_order_failed', 'Error when saving the order. Error: %s'),
```
Comment (fixpunkt) [inline engine/Shopware/Controllers/Backend/Order.php]: `when saving` -> `while trying to save`

File: engine/Shopware/Controllers/Backend/Order.php [hunk 2]
```diff
@@ -729,9 +754,22 @@ public function batchProcessAction()
             $data['paymentStatus'] = Shopware()->Models()->toArray($order->getPaymentStatus());
             $data['orderStatus'] = Shopware()->Models()->toArray($order->getOrderStatus());
 
-            $data['mail'] = $this->checkOrderStatus($order, $statusBefore, $clearedBefore, $autoSend, $documentType, $addAttachments);
-            //return the modified data array.
-            $orders[$key] = $data;
+            try {
+                // The method '$this->checkOrderStatus()' (even its name would not imply that) sends mails and can fail
```
Comment (fixpunkt) [inline engine/Shopware/Controllers/Backend/Order.php]: Please remove the passive-aggressive text in parentheses and re-wrap the comment 😉

File: snippets/backend/order/controller/main.ini [hunk 3]
```diff
@@ -0,0 +1,11 @@
+[en_GB]
+no_order_id_passed = "No valid order id passed."
+save_order_failed = "Error when saving the order. Error: %s"
```
Comment (fixpunkt) [inline snippets/backend/order/controller/main.ini]: `when saving` -> `while trying to save`

File: snippets/backend/order/controller/main.ini [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+send_mail_failed = "Error when sending mail. Error: %s"
```
Comment (fixpunkt) [inline snippets/backend/order/controller/main.ini]: `when sending mail` -> `while trying to send an email`

File: engine/Shopware/Controllers/Backend/Order.php [hunk 5]
```diff
@@ -729,9 +754,22 @@ public function batchProcessAction()
...
+                // with an exception. Catch this exception, so the batch process does not abort.
+                $data['mail'] = $this->checkOrderStatus($order, $statusBefore, $clearedBefore, $autoSend, $documentType, $addAttachments);
+            } catch (\Exception $e) {
+                $data['mail'] = null;
+                $data['success'] = false;
+                $data['errorMessage'] = sprintf(
+                    $namespace->get('send_mail_failed', 'Error when sending mail. Error: %s'),
```
Comment (fixpunkt) [inline engine/Shopware/Controllers/Backend/Order.php]: `when sending mail` -> `while trying to send an email`

File: snippets/backend/order/main.ini [hunk 6]
```diff
@@ -147,6 +147,7 @@ growlMessage = "Order"
 internal_comment = "Internal comment"
 layout = "View in"
 mail/title = "Send an email to the customer"
+batch/order_error/title = "Error at order [0]"
```
Comment (fixpunkt) [inline snippets/backend/order/main.ini]: `Error processing order [0]`

File: themes/Backend/ExtJs/backend/order/controller/batch.js [hunk 7]
```diff
@@ -70,6 +70,7 @@ Ext.define('Shopware.apps.Order.controller.Batch', {
             message: '{s name=cancel_message}{/s}',
             title: '{s name=cancel_title}{/s}'
         },
+        orderErrorTitle: '{s name=batch/order_error/title}Error at order [0]{/s}',
```
Comment (fixpunkt) [inline themes/Backend/ExtJs/backend/order/controller/batch.js]: `at` -> `processing`

File: themes/Backend/ExtJs/backend/order/controller/batch.js [hunk 8]
```diff
@@ -217,6 +218,17 @@ Ext.define('Shopware.apps.Order.controller.Batch', {
                     Shopware.Msg.createGrowlMessage('', me.snippets.noEmailsWillBeSent);
                 }
 
+                Ext.Array.forEach(resultSet, function (batchResult) {
```
Comment (fixpunkt) [inline themes/Backend/ExtJs/backend/order/controller/batch.js]: Please add a comment above this line explaining what happens here.

File: themes/Backend/ExtJs/backend/order/controller/batch.js [hunk 9]
```diff
@@ -501,6 +513,15 @@ Ext.define('Shopware.apps.Order.controller.Batch', {
         // Add the resulting record to our result store
         resultStore.add(operation.resultSet.records);
 
+        var resultRecord = operation.resultSet.records[0];
+        if (resultRecord.get('success') === false) {
```
Comment (fixpunkt) [inline themes/Backend/ExtJs/backend/order/controller/batch.js]: Is the `=== false` really necessary? `!resultRecord.get('success')` should be just fine too!

File: engine/Shopware/Controllers/Backend/Order.php [hunk 10]
```diff
@@ -441,12 +441,23 @@ public function saveAction()
...
-                $mail = $this->getMailForOrder($order->getId(), $order->getOrderStatus()->getId());
+                $status = $order->getOrderStatus();
             } else {
-                $mail = $this->getMailForOrder($order->getId(), $order->getPaymentStatus()->getId());
+                $status = $order->getPaymentStatus();
+            }
+            try {
+                $mail = $this->getMailForOrder($order->getId(), $status->getId());
+            } catch (\Exception $e) {
+                $mail = null;
```
Comment (fixpunkt) [inline engine/Shopware/Controllers/Backend/Order.php]: Redundant assignment.

File: engine/Shopware/Controllers/Backend/Order.php [hunk 11]
```diff
@@ -685,17 +701,21 @@ public function batchProcessAction()
             return;
         }
 
-        foreach ($orders as $key => $data) {
-            $orders[$key]['mail'] = null;
-            $orders[$key]['languageSubShop'] = null;
+        foreach ($orders as &$data) {
```
Comment (fixpunkt) [inline engine/Shopware/Controllers/Backend/Order.php]: Is it really necessary to change this loop to use references? I'd avoid this like hell tbh.

File: engine/Shopware/Controllers/Backend/Order.php [hunk 12]
```diff
@@ -685,17 +701,21 @@ public function batchProcessAction()
...
+            $data['mail'] = null;
+            $data['languageSubShop'] = null;
 
-            if (empty($data) || empty($data['id'])) {
+            if (empty($data['id'])) {
+                $data['success'] = false;
+                $data['errorMessage'] = $namespace->get('no_order_id_passed', 'No valid order id passed.');
```
Comment (fixpunkt) [inline engine/Shopware/Controllers/Backend/Order.php]: Is this really an error we need to report back to the user? From reading the code, this looks like a sanity check but shouldn't really happen in production?

File: engine/Shopware/Controllers/Backend/Order.php [hunk 13]
```diff
@@ -685,17 +701,21 @@ public function batchProcessAction()
...
                 continue;
             }
 
             /** @var $order \Shopware\Models\Order\Order */
             $order = Shopware()->Models()->find(Order::class, $data['id']);
             if (!$order) {
+                $data['success'] = false;
+                $data['errorMessage'] = $namespace->get('no_order_id_passed', 'No valid order id passed.');
```
Comment (fixpunkt) [inline engine/Shopware/Controllers/Backend/Order.php]: Same here.

**Review by fixpunkt — CHANGES_REQUESTED** [review]
Nice work!

I've updated the title and the description.

Please remember to always change the PR title from the default value (i.e. the branch name) to something more descriptive when opening PRs.

**Comment by windaishi** [comment]
Actually you reviewed the wrong PR :S. The original PR this PR is based on has already been openen a decade ago.

**Review by fixpunkt — APPROVED** [review]
Looks good!

**Comment by windaishi** [comment]
I am going to merge this PR into this one: https://github.com/shopware/shopware/pull/1496

**Pull request merged**